griptape/tools/google_drive/tool.py

```diff
@@ -0,0 +1,293 @@
+
+            while True:
+                results = service.files().list(
+                    q=query,
+                    pageSize=max_files,
+                    pageToken=next_page_token,
+                ).execute()
+
+                files = results.get('files', [])
+                items.extend(files)
+
+                next_page_token = results.get('nextPageToken')
+
+                if not next_page_token:
+                    break
+
+            return ListArtifact(items)
+
+        except Exception as e:
+            logging.error(e)
+            return ErrorArtifact(f"error retrieving files from Google Drive: {e}")
+
+    @activity(config={
+        "description": "Can be used to save memory artifacts to Google Drive using folder paths",
+        "schema": Schema({
+            "memory_name": str,
+            "artifact_namespace": str,
+            "file_name": str,
+            Optional("folder_path",
+                     description="Path of the Google Drive folder (like 'MainFolder/Subfolder1/Subfolder2') "
+                                 "where the file should be stored. If not provided, the file will be saved "
+                                 "to the root directory."): str,
+        })
+    })
+    def save_memory_artifacts_to_drive(self, params: dict) -> ErrorArtifact | InfoArtifact:
+        values = params["values"]
+        memory = self.find_input_memory(values["memory_name"])
+        file_name = values["file_name"]
+    
+        if memory:
+            artifacts = memory.load_artifacts(values["artifact_namespace"])
+    
+            if not artifacts:
+                return ErrorArtifact("no artifacts found")
+
+            service = self._build_client(self.UPLOAD_FILE_SCOPES)
+
+            folder_id = None
+            if "folder_path" in values and values["folder_path"]:
+                folder_id = self._path_to_file_id(service, values["folder_path"])
+                if not folder_id:
+                    return ErrorArtifact(f"Could not find folder: {values['folder_path']}")
+
+            try:
+                if len(artifacts) == 1:
+                    self._save_to_drive(file_name, artifacts[0].value, folder_id)
+                else:
+                    for a in artifacts:
+                        self._save_to_drive(f"{a.name}-{file_name}", a.value, folder_id)
+    
+                return InfoArtifact(f"saved successfully")
+    
+            except Exception as e:
+                return ErrorArtifact(f"error saving file to Google Drive: {e}")
+        else:
+            return ErrorArtifact("memory not found")
+
+    @activity(config={
+        "description": "Can be used to save content to a file",
+        "schema": Schema(
+            {
+                Literal(
+                    "path",
+                    description="Destination file path on disk in the POSIX format. For example, 'foo/bar/baz.txt'"
+                ): str,
+                "content": str
+            }
+        )
```

Comment:
Doesn't look like you modified those after copying them from `FileManager`.

Reply:
Done!